src/targets/cpu/binary.cpp

```diff
@@ -49,9 +49,21 @@ struct dnnl_binary : dnnl_op<dnnl_binary, dnnl::binary>
         auto s0 = inputs.at(0);
         auto s1 = inputs.at(1);
         auto r  = s0;
-        if(s0 != s1 or not s0.packed())
+        if(s0 == s1 and s0.packed())
         {
-            r = shape{s0.type(), s0.lens()};
+            r = s0;
+        }
+        else if(s0.packed() != s1.packed())
+        {
+            r = s0.packed() ? s0 : s1;
+        }
+        else if(s0.broadcasted() != s1.broadcasted())
+        {
+            r = s0.broadcasted() ? s1.with_lens(s0.lens()) : s0.with_lens(s0.lens());
+        }
+        else
+        {
+            r = {s0.type(), s0.lens()};
```

Comment:
We want to use the permutation of the packed shape, if neither is packed then try to pick the permutation of the non-broadcasted shape otherwise we just return a standard shape.

Reply:
Ah I see.